src/Storages/MergeTree/MergeTreeWriteAheadLog.cpp

```diff
@@ -0,0 +1,193 @@
+    extern const int BAD_DATA_PART_NAME;
+    extern const int CORRUPTED_DATA;
+}
+
+
+MergeTreeWriteAheadLog::MergeTreeWriteAheadLog(
+    const MergeTreeData & storage_,
+    const DiskPtr & disk_,
+    const String & name_)
+    : storage(storage_)
+    , disk(disk_)
+    , name(name_)
+    , path(storage.getRelativeDataPath() + name_)
+{
+    init();
+}
+
+void MergeTreeWriteAheadLog::init()
+{
+    out = disk->writeFile(path, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append);
+    block_out = std::make_unique<NativeBlockOutputStream>(*out, 0, storage.getSampleBlock());
+    min_block_number = std::numeric_limits<Int64>::max();
+    max_block_number = -1;
+}
+
+void MergeTreeWriteAheadLog::addPart(const Block & block, const String & part_name)
+{
+    std::lock_guard lock(write_mutex);
+
+    auto part_info = MergeTreePartInfo::fromPartName(part_name, storage.format_version);
+    min_block_number = std::min(min_block_number, part_info.min_block);
+    max_block_number = std::max(max_block_number, part_info.max_block);
+
+    writeIntBinary(static_cast<UInt8>(0), *out); /// version
+    writeIntBinary(static_cast<UInt8>(ActionType::ADD_PART), *out);
+    writeStringBinary(part_name, *out);
+    block_out->write(block);
+    block_out->flush();
+
+    auto max_wal_bytes = storage.getSettings()->write_ahead_log_max_bytes;
+    if (out->count() > max_wal_bytes)
+        rotate();
+}
+
+void MergeTreeWriteAheadLog::dropPart(const String & part_name)
+{
+    std::lock_guard lock(write_mutex);
+
+    writeIntBinary(static_cast<UInt8>(0), *out);
+    writeIntBinary(static_cast<UInt8>(ActionType::DROP_PART), *out);
+    writeStringBinary(part_name, *out);
+}
+
+void MergeTreeWriteAheadLog::rotate()
+{
+    String new_name = String(WAL_FILE_NAME) + "_"
+        + toString(min_block_number) + "_"
+        + toString(max_block_number) + WAL_FILE_EXTENSION;
+
+    disk->replaceFile(path, storage.getRelativeDataPath() + new_name);
+    init();
+}
+
+MergeTreeData::MutableDataPartsVector MergeTreeWriteAheadLog::restore()
+{
+    std::lock_guard lock(write_mutex);
+
+    MergeTreeData::MutableDataPartsVector parts;
+    auto in = disk->readFile(path, DBMS_DEFAULT_BUFFER_SIZE);
+    NativeBlockInputStream block_in(*in, 0);
+    NameSet dropped_parts;
+
+    while (!in->eof())
+    {
+        MergeTreeData::MutableDataPartPtr part;
+        UInt8 version;
+        String part_name;
+        Block block;
+        ActionType action_type;
+
+        try
+        {
+            readIntBinary(version, *in);
+            if (version != 0)
+                throw Exception("Unknown WAL format version: " + toString(version), ErrorCodes::UNKNOWN_FORMAT_VERSION);
+
+            readIntBinary(action_type, *in);
+            readStringBinary(part_name, *in);
+
+            if (action_type == ActionType::DROP_PART)
+            {
+                dropped_parts.insert(part_name);
+            }
+            else if (action_type == ActionType::ADD_PART)
```

Comment:
Is it necessary to restore dropped parts? Maybe we should check dropped_parts set here and skip some bytes in our stream?

Reply:
The main trouble is that `ADD_PART` appears before `DROP_PART`. It's possible to get first only dropped parts, but it requires two looks on wal file and another special format to have possibility to skip bytes.